File: MiniDOMOffice/Office/Mail/Folders/MailFoldersCollection.cs
using System;
using System.Collections;
using System.Collections.Generic;
using DMD;
using DMD.XML;
using DMD.Databases;
using DMD.Databases.Collections;
using minidom;
using minidom.repositories;
using static minidom.Sistema;
using static minidom.Anagrafica;
using static minidom.Office;


namespace minidom
{
    public partial class Office
    {

        /// <summary>
        /// Collezione di <see cref="MailFolder"/>
        /// </summary>
        [Serializable]
        public class MailFoldersCollection
            : CCollection<MailFolder>
        {
            [NonSerialized] private MailApplication m_Application;
            [NonSerialized] private CUser m_User;
            [NonSerialized] private MailFolder m_Parent;


            /// <summary>
            /// Costruttore
            /// </summary>
            public MailFoldersCollection()
            {
                m_Application = null;
                m_User = null;
                m_Parent = null;
            }

            /// <summary>
            /// Costruttore
            /// </summary>
            /// <param name="parent"></param>
            public MailFoldersCollection(MailFolder parent) : this()
            {
                if (parent is null)
                    throw new ArgumentNullException("parent");
                Load(parent);
            }

            /// <summary>
            /// Costruttore
            /// </summary>
            /// <param name="user"></param>
            public MailFoldersCollection(Sistema.CUser user) : this()
            {
                if (user is null)
                    throw new ArgumentNullException("user");
                Load(user);
            }

            /// <summary>
            /// OnInsert
            /// </summary>
            /// <param name="index"></param>
            /// <param name="value"></param>
            protected override void OnInsert(int index, MailFolder value)
            {
                if (m_Application is object)
                    value.SetApplication(m_Application);
                if (m_User is object)
                    value.SetUtente(m_User);
                if (m_Parent is object)
                    value.SetParent(m_Parent);
                base.OnInsert(index, value);
            }

            /// <summary>
            /// OnSet
            /// </summary>
            /// <param name="index"></param>
            /// <param name="oldValue"></param>
            /// <param name="newValue"></param>
            protected override void OnSet(int index, MailFolder oldValue, MailFolder newValue)
            {
                if (m_Application is object)
                    newValue.SetApplication(m_Application);
                if (m_User is object)
                    newValue.SetUtente(m_User);
                if (m_Parent is object)
                    newValue.SetParent(m_Parent);
                base.OnSet(index, oldValue, newValue);
            }

            /// <summary>
            /// Carica 
            /// </summary>
            /// <param name="parent"></param>
            protected internal void Load(MailFolder parent)
            {
                this.m_Parent = parent;
                this.m_User = parent.Utente;
                this.SetApplication(parent.Application);
                this.Clear();
                using (var cursor = new MailFolderCursor())
                {
                    cursor.ParentID.Value = DBUtils.GetID(m_Parent, 0);
                    cursor.Stato.Value = ObjectStatus.OBJECT_VALID;
                    cursor.IgnoreRights = true;
                    cursor.ID.SortOrder = SortEnum.SORT_ASC;
                    while (cursor.Read())
                    {
                        Add(cursor.Item);
                    }
                }
            }

            /// <summary>
            /// Carica
            /// </summary>
            /// <param name="user"></param>
            protected internal void Load(CUser user)
            {
                this.Clear();
                this.m_User = user;
                using (var cursor = new MailFolderCursor())
                { 
                    cursor.Stato.Value = ObjectStatus.OBJECT_VALID;
                    cursor.IgnoreRights = true;
                    cursor.ParentID.Value = 0;
                    cursor.ParentID.IncludeNulls = true;
                    cursor.IDutente.Value = DBUtils.GetID(user, 0);
                    while (cursor.Read())
                    {
                        Add(cursor.Item);
                    }
                }
            }

            /// <summary>
            /// Restituisce la sottocartella con nome
            /// </summary>
            /// <param name="path"></param>
            /// <returns></returns>
            public MailFolder GetItemByName(string path)
            {
                path = Strings.Replace(Strings.Trim(path), "/", @"\");
                if (string.IsNullOrEmpty(path))
                    return null;
                int p = Strings.InStr(path, @"\");
                string name = "";
                if (p > 0)
                {
                    name = Strings.Trim(Strings.Left(path, p - 1));
                    path = Strings.Trim(Strings.Mid(path, p + 1));
                }
                else
                {
                    name = path;
                }

                if (string.IsNullOrEmpty(name))
                    return m_Application.Root.Childs.GetItemByName(path);
                if (name == "..")
                {
                    if (m_Parent is null)
                        return null;
                    return m_Parent.Childs.GetItemByName(path);
                }

                MailFolder ret = null;
                foreach (MailFolder f in this)
                {
                    if ((f.Name ?? "") == (name ?? ""))
                        return f;
                }

                if (ret is null)
                    return null;
                if (string.IsNullOrEmpty(path))
                    return ret;
                return ret.Childs.GetItemByName(path);
            }

            /// <summary>
            /// Aggiunge una nuova cartella con nome
            /// </summary>
            /// <param name="folderName"></param>
            /// <returns></returns>
            public MailFolder Add(string folderName)
            {
                folderName = DMD.Strings.Trim(folderName);
                if (folderName.IndexOf("/") >= 0)
                    throw new ArgumentException("Il percorso non può contenere caratteri speciali");
                var f = new MailFolder();
                f.Name = folderName;
                f.Stato = ObjectStatus.OBJECT_VALID;
                Add(f);
                f.Save();
                return f;
            }

            /// <summary>
            /// Imposta l'utente
            /// </summary>
            /// <param name="user"></param>
            protected internal void SetUser(Sistema.CUser user)
            {
                m_User = user;
                if (user is object)
                {
                    foreach (MailFolder f in this)
                        f.SetUtente(user);
                }
            }

            /// <summary>
            /// Imposta il genitore
            /// </summary>
            /// <param name="p"></param>
            protected internal void SetParent(MailFolder p)
            {
                m_Parent = p;
                if (p is object)
                {
                    foreach (MailFolder f in this)
                        f.SetParent(p);
                }
            }

            /// <summary>
            /// Imposta l'app
            /// </summary>
            /// <param name="a"></param>
            protected internal void SetApplication(MailApplication a)
            {
                m_Application = a;
                if (a is object)
                {
                    foreach (MailFolder f in this)
                        f.SetApplication(a);
                }
            }
        }
    }
}

File: MiniDOMOffice/Office/Mail/Folders/MailFoldersCollection.vb
Imports System
Imports System.Collections.Generic
Imports System.Text
Imports System.Text.RegularExpressions
Imports System.Net.Mail
Imports System.Net.Mime
Imports minidom.Net.Mime
Imports minidom.Databases
Imports minidom.Sistema

Partial Class Office

    <Serializable> _
    Public Class MailFoldersCollection
        Inherits CCollection(Of MailFolder)

        <NonSerialized> Private m_Application As MailApplication
        <NonSerialized> Private m_User As CUser
        <NonSerialized> Private m_Parent As MailFolder

        Public Sub New()
            Me.m_Application = Nothing
            Me.m_User = Nothing
            Me.m_Parent = Nothing
        End Sub

        Public Sub New(ByVal parent As MailFolder)
            Me.New()
            If (parent Is Nothing) Then Throw New ArgumentNullException("parent")
            Me.Load(parent)
        End Sub

        Public Sub New(ByVal user As CUser)
            Me.New()
            If (user Is Nothing) Then Throw New ArgumentNullException("user")
            Me.Load(user)
        End Sub

        Protected Overrides Sub OnInsert(index As Integer, value As Object)
            If (Me.m_Application IsNot Nothing) Then DirectCast(value, MailFolder).SetApplication(Me.m_Application)
            If (Me.m_User IsNot Nothing) Then DirectCast(value, MailFolder).SetUtente(Me.m_User)
            If (Me.m_Parent IsNot Nothing) Then DirectCast(value, MailFolder).SetParent(Me.m_Parent)
            MyBase.OnInsert(index, value)
        End Sub

        Protected Overrides Sub OnSet(index As Integer, oldValue As Object, newValue As Object)
            If (Me.m_Application IsNot Nothing) Then DirectCast(newValue, MailFolder).SetApplication(Me.m_Application)
            If (Me.m_User IsNot Nothing) Then DirectCast(newValue, MailFolder).SetUtente(Me.m_User)
            If (Me.m_Parent IsNot Nothing) Then DirectCast(newValue, MailFolder).SetParent(Me.m_Parent)
            MyBase.OnSet(index, oldValue, newValue)
        End Sub

        Protected Friend Sub Load(ByVal parent As MailFolder)
            Dim cursor As MailFolderCursor = Nothing
            Try
                Me.m_Parent = parent
                Me.m_User = parent.Utente
                Me.SetApplication(parent.Application)
                Me.Clear()
                If (GetID(Me.m_Parent) = 0) Then Exit Sub
                cursor = New MailFolderCursor
                cursor.ParentID.Value = GetID(Me.m_Parent)
                cursor.Stato.Value = ObjectStatus.OBJECT_VALID
                cursor.IgnoreRights = True
                cursor.ID.SortOrder = SortEnum.SORT_ASC
                While Not cursor.EOF
                    MyBase.Add(cursor.Item)
                    cursor.MoveNext()
                End While

            Catch ex As Exception
                Sistema.Events.NotifyUnhandledException(ex)
                Throw
            Finally
                If (cursor IsNot Nothing) Then cursor.Dispose() : cursor = Nothing
            End Try
        End Sub

        Protected Friend Sub Load(ByVal user As CUser)
            Dim cursor As MailFolderCursor = Nothing
            Try
                Me.Clear()
                Me.m_User = user
                If (GetID(user) = 0) Then Exit Sub
                cursor = New MailFolderCursor
                cursor.Stato.Value = ObjectStatus.OBJECT_VALID
                cursor.IgnoreRights = True
                cursor.ParentID.Value = 0
                cursor.ParentID.IncludeNulls = True
                cursor.IDutente.Value = GetID(user)
                While Not cursor.EOF
                    MyBase.Add(cursor.Item)
                    cursor.MoveNext()
                End While
            Catch ex As Exception
                Sistema.Events.NotifyUnhandledException(ex)
                Throw
            Finally
                If (cursor IsNot Nothing) Then cursor.Dispose() : cursor = Nothing
            End Try
        End Sub

        Public Function GetItemByName(ByVal path As String) As MailFolder
            path = Replace(Trim(path), "/", "\")
            If (path = "") Then Return Nothing
            Dim p As Integer = InStr(path, "\")
            Dim name As String = ""
            If (p > 0) Then
                name = Trim(Left(path, p - 1))
                path = Trim(Mid(path, p + 1))
            Else
                name = path
            End If
            If (name = "") Then Return Me.m_Application.Root.Childs.GetItemByName(path)
            If (name = "..") Then
                If (Me.m_Parent Is Nothing) Then Return Nothing
                Return Me.m_Parent.Childs.GetItemByName(path)
            End If
            Dim ret As MailFolder = Nothing
            For Each f As MailFolder In Me
                If (f.Name = name) Then Return f
            Next
            If (ret Is Nothing) Then Return Nothing
            If (path = "") Then Return ret
            Return ret.Childs.GetItemByName(path)
        End Function

        Public Overloads Function Add(ByVal folderName As String) As MailFolder
            folderName = Strings.Trim(folderName)
            If (folderName.IndexOf("/") >= 0) Then Throw New ArgumentException("Il percorso non può contenere caratteri speciali")
            Dim f As New MailFolder
            f.Name = folderName
            f.Stato = ObjectStatus.OBJECT_VALID
            Me.Add(f)
            f.Save()
            Return f
        End Function

        Protected Friend Sub SetUser(ByVal user As CUser)
            Me.m_User = user
            If (user IsNot Nothing) Then
                For Each f As MailFolder In Me
                    f.SetUtente(user)
                Next
            End If
        End Sub

        Protected Friend Sub SetParent(ByVal p As MailFolder)
            Me.m_Parent = p
            If (p IsNot Nothing) Then
                For Each f As MailFolder In Me
                    f.SetParent(p)
                Next
            End If
        End Sub

        Protected Friend Sub SetApplication(ByVal a As MailApplication)
            Me.m_Application = a
            If (a IsNot Nothing) Then
                For Each f As MailFolder In Me
                    f.SetApplication(a)
                Next
            End If
        End Sub

    End Class

End Class